File: integration-tests/src/main/java/ee/veikokaap/debugproxy/testframework/AsyncTester.java
package ee.veikokaap.debugproxy.testframework;

import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Consumer;

public class AsyncTester<T> implements Tester<T> {

  private final AtomicReference<Throwable> exception = new AtomicReference<>(null);
  private final AtomicReference<Thread> thread = new AtomicReference<>(null);
  private final Consumer<T> consumer;

  public AsyncTester(Consumer<T> consumer) {
    this.consumer = consumer;
  }

  @Override
  public void accept(T t) {
    if (exception.get() != null) {
      return; // Test previously failed
    }

    thread.set(Thread.currentThread());
    try {
      consumer.accept(t);
    }
    catch (AssertionError e) {
      exception.set(e);
    }
  }

  @Override
  public void joinAndTest(long time, TimeUnit timeUnit) throws Throwable {
    long millis = timeUnit.toMillis(time);
    long start = System.currentTimeMillis();

    waitForConsumerRun(millis, start);
    if (thread.get() == null) {
      throw new AssertionError("Consumer hasn't been run in " + time + " " + timeUnit);
    }
    thread.get().join(timeLeft(millis, start));

    if (thread.get().isAlive()) {
      thread.get().interrupt();
      TimeoutException timeoutException = new TimeoutException("Thread should have finished work in " + time + " " + timeUnit);
      if (exception.get() != null) {
        timeoutException.addSuppressed(exception.get());
      }
      throw timeoutException;
    }
    if (exception.get() != null) {
      throw exception.get();
    }
  }

  private long timeLeft(long millis, long start) {
    return millis - (System.currentTimeMillis() - start);
  }

  private void waitForConsumerRun(long millis, long start) throws InterruptedException {
    while (thread.get() == null && (System.currentTimeMillis() - start < millis)) {
      Thread.sleep(100);
    }
  }

  public void failTestWithException(Exception e) {
    exception.set(e);
  }
}

File: integration-tests/src/main/kotlin/ee/veikokaap/debugproxy/testframework/AsyncTester.kt
package ee.veikokaap.debugproxy.testframework

import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import java.util.concurrent.atomic.AtomicReference
import java.util.function.Consumer

open class AsyncTester<T>(private val consumer: Consumer<T>): Consumer<T> {

    protected val exception = AtomicReference<Throwable>(null)
    private val thread = AtomicReference<Thread>(null)

    override fun accept(t: T) {
        if (exception.get() != null) {
            return  // Test previously failed
        }

        thread.set(Thread.currentThread())
        try {
            consumer.accept(t)
        } catch (e: Throwable) {
            exception.set(e)
        }
    }

    @Throws(Throwable::class)
    fun joinAndTest() {
        joinAndTest(Long.MAX_VALUE, TimeUnit.DAYS);
    }

    @Throws(Throwable::class)
    fun joinAndTest(time: Long, timeUnit: TimeUnit) {
        val millis = timeUnit.toMillis(time)
        val start = System.currentTimeMillis()

        waitForConsumerRun(millis, start)
        if (thread.get() == null) {
            throw AssertionError("Consumer hasn't been run in $time $timeUnit")
        }

        val timeLeft = timeLeft(millis, start)
        if (timeLeft <= 0) {
            throwTimeoutException(time, timeUnit)
        }
        thread.get().join(timeLeft)

        if (thread.get().isAlive) {
            throwTimeoutException(time, timeUnit)
        }
        if (exception.get() != null) {
            throw exception.get()
        }
    }

    private fun throwTimeoutException(time: Long, timeUnit: TimeUnit) {
        thread.get().interrupt()
        val timeoutException = TimeoutException("Thread should have finished work in $time $timeUnit")
        if (exception.get() != null) {
            timeoutException.addSuppressed(exception.get())
        }
        throw timeoutException
    }

    private fun timeLeft(millis: Long, start: Long) = millis - (System.currentTimeMillis() - start)

    @Throws(InterruptedException::class)
    private fun waitForConsumerRun(millis: Long, start: Long) {
        while (thread.get() == null && System.currentTimeMillis() - start < millis) {
            Thread.sleep(100)
        }
    }

    fun failTestWithException(e: Exception) = exception.set(e)

}
